sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -270,6 +273,146 @@ class Analyzer(
   }
 
   /**
+   * Rewrite the [[Exists]] [[In]] with left semi join or anti join.
+   */
+  object RewriteFilterSubQuery extends Rule[LogicalPlan] with PredicateHelper {
+    def unapply(condition: Expression): Option[(Expression, Seq[Expression])] = {
+      if (condition.resolved == false) {
+        return None
+      }
+
+      val conjuctions = splitConjunctivePredicates(condition).map(_ transformDown {
+          // Remove the Cast expression for SubQueryExpression.
+          case Cast(f: SubQueryExpression, BooleanType) => f
+        }
+      )
+
+      val (subqueries, others) = conjuctions.partition(c => c.isInstanceOf[SubQueryExpression])
+      if (subqueries.isEmpty) {
+        None
+      } else if (subqueries.length > 1) {
+        throw new AnalysisException(
+          s"Only 1 SubQuery expression is supported, but we got $subqueries")
+      } else {
+        val subQueryExpr = subqueries(0).asInstanceOf[SubQueryExpression]
+        // try to resolve the subquery
+
+        val subquery = Analyzer.this.execute(subQueryExpr.subquery) match {
+          case Distinct(child) => child // Distinct is useless for semi join, ignore it.
+          case other => other
+        }
+        Some((subQueryExpr.withNewSubQuery(subquery), others))
+      }
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+      case f if f.childrenResolved == false => f
+
+      case f @ Filter(RewriteFilterSubQuery(subquery, others), left) =>
+        subquery match {
+          case Exists(Project(_, Filter(condition, right)), positive) =>
```

Comment:
This is the only case we support right now for `EXISTS(NOT)`. (This is the correlated case, the correlated reference via the `Filter` operator)
`SELECT value FROM src a WHERE EXISTS (SELECT key FROM src b WHERE a.key=b.key AND a.key> 100)`.

(These are the unrelated cases)
`SELECT value FROM src a WHERE EXISTS (SELECT key FROM src b)`. // without Filter
`SELECT value FROM src a WHERE EXISTS (SELECT key FROM src b WHERE a.key> 100)`. // Filter condition is resolved, then definitely an unrelated reference.